source/extensions/transport_sockets/alts/config.cc

```diff
@@ -0,0 +1,144 @@
+#include "extensions/transport_sockets/alts/tsi_socket.h"
+
+#include "absl/strings/str_join.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+typedef CSmartPtr<grpc_alts_credentials_options, grpc_alts_credentials_options_destroy>
+    GrpcAltsCredentialsOptionsPtr;
+
+// Returns true if the peer's service account is found in peers, otherwise
+// returns false and fills out err with an error message.
+static bool doValidate(const tsi_peer& peer, const std::unordered_set<std::string>& peers,
+                       std::string& err) {
+  for (size_t i = 0; i < peer.property_count; ++i) {
+    std::string name = std::string(peer.properties[i].name);
+    std::string value = std::string(peer.properties[i].value.data, peer.properties[i].value.length);
+    if (name.compare(TSI_ALTS_SERVICE_ACCOUNT_PEER_PROPERTY) == 0 &&
+        peers.find(value) != peers.end()) {
+      return true;
+    }
+  }
+
+  err =
+      "Couldn't find peer's service account in peer_service_accounts: " + absl::StrJoin(peers, ",");
+  return false;
+}
+
+ProtobufTypes::MessagePtr AltsTransportSocketConfigFactory::createEmptyConfigProto() {
+
+  return std::make_unique<envoy::config::transport_socket::alts::v2alpha::Alts>();
+}
+
+Network::TransportSocketFactoryPtr
+UpstreamAltsTransportSocketConfigFactory::createTransportSocketFactory(
+    const Protobuf::Message& message, Server::Configuration::TransportSocketFactoryContext&) {
+  auto config =
+      MessageUtil::downcastAndValidate<const envoy::config::transport_socket::alts::v2alpha::Alts&>(
+          message);
+
+  std::string handshaker_service = config.handshaker_service();
+  const auto& peer_service_accounts = config.peer_service_accounts();
+  std::unordered_set<std::string> peers(peer_service_accounts.cbegin(),
+                                        peer_service_accounts.cend());
+
+  HandshakeValidator validator;
+  // Skip validation if peers is empty.
+  if (!peers.empty()) {
+    validator = [peers](const tsi_peer& peer, std::string& err) {
+      return doValidate(peer, peers, err);
+    };
+  }
+
+  return std::make_unique<TsiSocketFactory>(
+      [handshaker_service](Event::Dispatcher& dispatcher,
+                           const Network::Address::InstanceConstSharedPtr&,
+                           const Network::Address::InstanceConstSharedPtr&) -> TsiHandshakerPtr {
+        GrpcAltsCredentialsOptionsPtr options{grpc_alts_credentials_client_options_create()};
+
+        tsi_handshaker* handshaker = nullptr;
+        // Specifying target name as empty since TSI won't take care of validating peer identity
+        // in this use case. The validation will be performed by TsiSocket with the validator.
+        tsi_result status = alts_tsi_handshaker_create(
+            options.get(), "", handshaker_service.c_str(), true /* is_client */, &handshaker);
+        CHandshakerPtr handshaker_ptr{handshaker};
```

Comment:
Should this be moved to after the status check below (for clarity, rather than correctness)?